lib/include/Tensile/Singleton.hpp

```diff
@@ -0,0 +1,61 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2019 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <mutex>
+
+namespace Tensile
+{
+    template <typename Class>
+    class LazySingleton
+    {
+    public:
+        static Class & Instance()
+        {
+            std::call_once(m_instanceConstructionFlag, Construct);
+
+            return *m_instance;
+        }
+
```

Comment:
Much simpler, without a need for `Construct()`, external `m_instance`, or `m_instanceConstructionFlag`:

    static auto& Instance()
    {
       static auto* m_instance = new Class();
       return *m_instance;
    }

Reply:
Like above, you might not need dynamic allocation:

    static auto& Instance()
    {
        static Class m_instance;
        return m_instance;
    }

Reply:
I guess that's a question of if we want the objects to be instantiated at static initialization time, or at the time of first use.  Changing that behaviour would make this a StaticSingleton, not a LazySingleton.  We could have both.

Reply:
The codes which used `call_once` were initializing their important data (not the `call_once_flag`) at time of first use; I'm merely suggesting that it be implemented using local `static` variables instead.

Initialization at startup is generally bad because initialization order depends on link or load order. With local `static` initialization occurs at time of first use and does not suffer from indeterminate initialization order.

`call_once` is slower than doing the equivalent with local `static` variables.

Reply:
Done.